File: NinePSharp/messages/Rwalk.cs
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Rwalk tag[2] nwqid[2] nwqid*(wqid[13])
public readonly struct Rwalk : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rwalk;
    public ushort Tag { get; }

    public Qid[] Wqid { get; }

    public Rwalk(ushort tag, Qid[] wqid)
    {
        Tag = tag;
        Wqid = wqid ?? Array.Empty<Qid>();
        Size = (uint)(NinePConstants.HeaderSize + 2 + (Wqid.Length * 13));
    }

    public Rwalk(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));

        int offset = NinePConstants.HeaderSize;

        ushort nwqid = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(offset, 2));
        offset += 2;

        Wqid = new Qid[nwqid];
        for (int i = 0; i < nwqid; i++)
        {
            Wqid[i] = data.ReadQid(ref offset);
        }
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;

        ushort nwqid = (ushort)(Wqid?.Length ?? 0);
        BinaryPrimitives.WriteUInt16LittleEndian(data.Slice(offset, 2), nwqid);
        offset += 2;

        for (int i = 0; i < nwqid; i++)
        {
            data.WriteQid(Wqid![i], ref offset);
        }
    }
}

File: NinePSharp.Messages.FSharp/messages/Rwalk.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rwalk =
    { Size: uint32
      Tag: uint16
      Wqid: Qid[] }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rwalk
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rwalk span &offset
            let len = match this.Wqid with | null -> 0us | x -> uint16 x.Length
            BinaryPrimitives.WriteUInt16LittleEndian(span.Slice(offset, 2), len); offset <- offset + 2
            if not (isNull this.Wqid) then
                for qid in this.Wqid do
                    writeQid qid span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let nwqid = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(offset, 2))
        offset <- offset + 2
        let wqid = Array.zeroCreate (int nwqid)
        for i = 0 to int nwqid - 1 do
            wqid.[i] <- readQid data &offset
        { Size = size; Tag = tag; Wqid = wqid }

